core/src/main/scala/org/apache/spark/shuffle/hash/HashShuffleWriter.scala

```diff
@@ -106,6 +108,19 @@ private[spark] class HashShuffleWriter[K, V](
       writer.commitAndClose()
       writer.fileSegment().length
     }
+    // rename all shuffle files to final paths
+    shuffle.writers.zip(sizes).foreach { case (writer: DiskBlockObjectWriter, size: Long) =>
+      if (size > 0) {
+        val output = blockManager.diskBlockManager.getFile(writer.blockId)
+        if (output.exists()) {
+          writer.file.delete()
+        } else {
+          if (!writer.file.renameTo(output)) {
+            throw new IOException(s"fail to rename ${writer.file} to $output")
```

Comment:
same problem here on partially existing shuffle output.  Also, the `if (size > 0)` check will lead to inconsistencies if you have non-deterministic shuffle output -- it might be true for different partitions in different attempts.  I think it can be OK in any case, as long as the MapStatus is consistent.